bson-bench/src/common.ts

```diff
@@ -0,0 +1,199 @@
+// handle local package
+export const LOCAL_PACKAGE_REGEX = /(bson-ext|bson):(.+)/;
+
+/**
+ * The Package class represents the bson or bson-ext package to be tested
+ * This package can be an npm package, a git repository or a local package
+ **/
+export class Package {
+  type: "npm" | "git" | "local";
+  // bson library to install
+  library: "bson" | "bson-ext";
+  computedModuleName: string;
+  // semver version specification
+  npmVersion?: string;
+  // git commit hash or tag
+  gitCommitish?: string;
+  // path to local library
+  localPath?: string;
+
+  constructor(libSpec: string) {
+    let match: RegExpExecArray | null;
+    if ((match = NPM_PACKAGE_REGEX.exec(libSpec))) {
+      this.type = "npm";
+      this.library = match[1] as "bson" | "bson-ext";
+      this.npmVersion = match[2];
+      this.computedModuleName = `${this.library}-${this.npmVersion}`;
+    } else if ((match = GIT_PACKAGE_REGEX.exec(libSpec))) {
+      this.type = "git";
+      this.library = match[1] as "bson" | "bson-ext";
+      this.gitCommitish = match[2];
+      this.computedModuleName = `${this.library}-git-${this.gitCommitish}`;
+    } else if ((match = LOCAL_PACKAGE_REGEX.exec(libSpec))) {
+      this.type = "local";
+      this.library = match[1] as "bson" | "bson-ext";
+
+      this.localPath = match[2];
+      this.computedModuleName = `${
+        this.library
+      }-local-${this.localPath.replaceAll(path.sep, "_")}`;
+    } else {
+      throw new Error("unknown package specifier");
+    }
+  }
+
+  /**
+   * returns the package if it exists, otherwise return undefined
+   */
+  check(): BSONLib | undefined {
+    try {
+      return require(this.computedModuleName);
+    } catch (e) {
+      return undefined;
+    }
+  }
+
+  /**
+   * Installs the packag e using the  n pm  cli
+   * This function is idempotent
+   *
+   * Note that this function cannot be run in the same Node process that imports the installed
+   * module
+   **/
+  async install(): Promise<void> {
+    let source: string;
+    switch (this.type) {
+      case "npm":
+        source = `npm:${this.library}@${this.npmVersion}`;
+        break;
+      case "git":
+        switch (this.library) {
+          case "bson":
+            source = `git://github.com/mongodb/js-bson#${this.gitCommitish}`;
+            break;
+          case "bson-ext":
+            source = `git://github.com/mongodb-js/bson-ext#${this.gitCommitish}`;
+            break;
+        }
+        break;
+      case "local":
+        source = `${this.localPath}`;
+        // Check if path exists as npm install will not throw an error if this is not the case
+        if ((await fs.access(source)) as unknown)
+          throw new Error(`'${source}' not found`);
+        break;
+    }
+
+    // TODO: Capture stdout and stderr
+    const pathToNpm = path.join(
+      __dirname,
+      "..",
+      "node_modules",
+      "npm",
+      "bin",
+      "npm-cli.js",
+    );
```

Comment:
Here we use the npm we added to the dependencies, so we don't depend on the user's npm installation.